### PR TITLE
feat(storage): add EmptyDir storage option

### LINKED ISSUE #220: Allow emptyDir volumes instead of PVCs
We should add an option in `Spec.StorageOptions` to deploy Cryostat using `emptyDir` volumes instead of creating a Persistent Volume Claim. If neither are specified, we can retain the same default of creating a PVC.

### PR DESCRIPTION
**Pull request by jan-law**
Fixes #220 

Adds a configuration option under `spec.storageOptions` to deploy Cryostat with `emptyDir` volumes with a default unbounded size limit. If an `emptyDir` configuration is specified, any `pvc.spec` configurations will be ignored.

```
apiVersion: operator.cryostat.io/v1beta1
kind: Cryostat
metadata:
  name: cryostat-sample
spec:
  storageOptions:
    emptyDir:
      enabled: true
      sizeLimit: 1Gi
```

### CODE REVIEW

**Review by ebaron — COMMENTED** [review]
Hey Janelle, this is looking good! I'm thinking perhaps the default capacity should be unbounded to match the Kubernetes defaults for emptyDir volumes [1].

I wonder if we need something extra to make this work nicely with the form view in OpenShift, when the user just wants a default empty dir. With YAML, they can do: `spec.storageOptions.emptyDir: {}`, but I don't think the user can trigger this empty object using the form view. Maybe we should add an `enabled` boolean to the EmptyDirConfig. You can even add dependencies on fields so the Medium and SizeLimit fields are only editable when `enabled` is true [2].

[1] https://kubernetes.io/docs/reference/kubernetes-api/config-and-storage-resources/volume/#local-temporary-directory
[2] https://github.com/openshift/console/blob/eadd6140e5dd5885f72b6f272da5eb8603914bd9/frontend/packages/operator-lifecycle-manager/src/components/descriptors/reference/reference.md#20-fielddependency

File: bundle/manifests/cryostat-operator.clusterserviceversion.yaml [hunk 1]
```diff
@@ -180,6 +180,24 @@ spec:
...
+      - description: Unless specified, the emptyDir volume will be mounted on the same storage medium backing the node. Setting this field to "Memory" will mount the emptyDir on a tmpfs (RAM-backed filesystem).
+        displayName: Medium
+        path: storageOptions.emptyDir.medium
+        x-descriptors:
+        - urn:alm:descriptor:com.tectonic.ui:fieldDependency:spec.storageOptions.emptyDir.enabled:true
+      - description: The maximum memory limit for the emptyDir. Default is unbounded.
+        displayName: Size Limit
+        path: storageOptions.emptyDir.sizeLimit
+        x-descriptors:
+        - urn:alm:descriptor:com.tectonic.ui:fieldDependency:spec.storageOptions.emptyDir.enabled:true
```
Comment (ebaron) [inline bundle/manifests/cryostat-operator.clusterserviceversion.yaml:200]: For the `fieldDependency` descriptors, the `spec` property is already implied, so these should begin with `storageOptions`.

There's unfortunately another problem that prevents `sizeLimit` from appearing in the OpenShift Console form view: https://github.com/openshift/console/issues/8298. `resource.Quantity` is treated as an int-or-string type by controller-tools, which is used by the Operator SDK: https://github.com/kubernetes-sigs/controller-tools/blob/97ee66cc83b03bd632cf8da35496812ca091ae4e/pkg/crd/known_types.go#L61-L70

We could work around this by changing `sizeLimit` to a `*string` and adding the validation manually with:
`// +kubebuilder:validation:Pattern=^(\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))(([KMGTPE]i)|[numkMGTPE]|([eE](\+|-)?(([0-9]+(\.[0-9]*)?)|(\.[0-9]+))))?$` and then parsing the string into a Quantity with `resource.ParseQuantity`. What do you think?

Comment (jan-law) [reply bundle/manifests/cryostat-operator.clusterserviceversion.yaml]: That's unfortunate about the `sizeLimit` form field. I've changed the field to a string and added the validation you suggested.

File: internal/controllers/common/resource_definitions/resource_definitions.go [hunk 2]
```diff
@@ -1205,3 +1198,37 @@ func getPullPolicy(imageTag string) corev1.PullPolicy {
 	// Likely a release, use IfNotPresent
 	return corev1.PullIfNotPresent
 }
+
+func newVolumeForCR(cr *operatorv1beta1.Cryostat) []corev1.Volume {
+	var volumeSource corev1.VolumeSource
+	deployEmptyDir := cr.Spec.StorageOptions != nil && cr.Spec.StorageOptions.EmptyDir != nil && cr.Spec.StorageOptions.EmptyDir.Enabled
```
[the remark this answers is not in the document]
Comment (ebaron) [reply internal/controllers/common/resource_definitions/resource_definitions.go:1204]: That's odd. It's working for me with `make deploy_bundle BUNDLE_IMG=quay.io/ebaron/cryostat-operator-bundle:emptydir-01`. Does that work for you?

[the remark this answers is not in the document]
Comment (jan-law) [reply internal/controllers/common/resource_definitions/resource_definitions.go:1204]: Hmm, I'll try it again once cluster-bot launches another cluster. Is this what you used to build your bundle?
```
make bundle bundle-build BUNDLE_IMG=quay.io/jalaw/cryostat-operator-bundle:latest
podman push quay.io/jalaw/cryostat-operator-bundle:latest
make undeploy_bundle
make deploy_bundle BUNDLE_IMG=quay.io/jalaw/cryostat-operator-bundle:latest
```

**Review by ebaron — APPROVED** [review]
Looks good Janelle. Thank you!

**Pull request merged**